keps/sig-api-machinery/2887-openapi-enum-types/README.md

```diff
@@ -0,0 +1,836 @@
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Currently, types that are actually enums are represented as plain strings. For example,
+here is the definition of `Protocol`, which must be one of `TCP`, `UDP`, and `SCTP`.
+```go
+package core
+// ...
+
+// Protocol defines network protocols supported for things like container ports.
+type Protocol string
+
+const (
+	// ProtocolTCP is the TCP protocol.
+	ProtocolTCP Protocol = "TCP"
+	// ProtocolUDP is the UDP protocol.
+	ProtocolUDP Protocol = "UDP"
+	// ProtocolSCTP is the SCTP protocol.
+	ProtocolSCTP Protocol = "SCTP"
+)
+
+// ContainerPort represents a network port in a single container
+type ContainerPort struct {
+  // ...
+
+  // Required: Supports "TCP", "UDP" and "SCTP"
+  // +optional
+  Protocol Protocol
+  
+  // ...
+}
+```
+
+In order to convey that the `Protocol` type is an enum, every field of the type 
+have to comment its values like `ContainerPort` does. The same goes for validation - 
+protocol being one of "TCP", "UDP", and "SCTP" is duplicated among the code base.
+
+Tools and UIs can also benefit from the additional piece of information. For example,
+Protocol can be shown as a dropdown of possible values instead of a textbox.
+The user no longer needs to see an error of invalid value after submits the request.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Add a marker that indicates a field as an enum.
+- Make the Kubernetes OpenAPI generator recognize the marker.
+- Deduce values of an enum type from its Go type definition.
+- Detect and annotate enum types in all built-in types. 
+- Unify the marker with that of kube-builder's enum mechanism.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+This KEP does not provide support for validation that is more complicated than an enum type
+like regex or prefix. Instead, use other validation mechanism that is or will be supported.
+
+This KEP does not consider extensibility of enum values. Any changes to the set of possible values
+should be considered breaking change and thus should follow the standard procedure of changing an API.
+
+## Proposal
+
+When generating OpenAPI schema, the de facto enum types should be reflected as enums in the resulting schema.
+
+We define enum types with following properties:
+- All enum types are closed, i.e., the type includes a finite number of possible values.
+- All enum types should be string. There can be no enum integers, floats, etc
+
+Additionally, This KEP makes the following assumptions on the implementation:
+- All enum types have their own type alias. i.e. , no raw string enums.
+- The canonical definition of a type is written in go and follows the standard 
```

Comment:
Maybe put this sentence first, otherwise it's not clear when you say "their own type alias" that you're referring to the go definition.

Reply:
true. updated/